activerecord/lib/active_record/associations.rb

```diff
@@ -574,7 +574,8 @@ def association_instance_set(name, association)
       #   end
       #
       # Note: Joining, eager loading and preloading of these associations is not possible.
-      # These operations happen before instance creation and the scope will be called with a +nil+ argument.
+      # These operations happen before instance creation, and the scope will be called with a +nil+ argument
+      # only if the scope proc's argument has a default value.
```

Comment:
Looking at #36134, I don't think supporting a default arg is intended.  Is there a use case where you would want the scope to be called with `nil`, instead of [raising an `ArgumentError`](https://github.com/rails/rails/blob/099289b6360ac82d1e0fa0a5592ac10cfc05e6e0/activerecord/lib/active_record/reflection.rb#L490-L496)?

Also, note that lambda and non-lambda Procs behave differently:

```console
> ->(x = 1){}.arity # lambda
-1

> Proc.new{|x = 1|}.arity # non-lambda
0
```

So perhaps we should just abridge the note:

```diff
diff --git a/activerecord/lib/active_record/associations.rb b/activerecord/lib/active_record/associations.rb
index 35a8aa2939..c7372f16bf 100644
--- a/activerecord/lib/active_record/associations.rb
+++ b/activerecord/lib/active_record/associations.rb
@@ -573,8 +573,8 @@ def association_instance_set(name, association)
       #     has_many :birthday_events, ->(user) { where(starts_on: user.birthday) }, class_name: 'Event'
       #   end
       #
-      # Note: Joining, eager loading and preloading of these associations is not possible.
-      # These operations happen before instance creation and the scope will be called with a +nil+ argument.
+      # Note: Joining, eager loading, and preloading of these associations is not possible
+      # because these operations happen before instance creation.
       #
       # == Association callbacks
       #
```

Reply:
I believe the docs are correct, just incomplete. It is useful to be able to define an association with a scope that can be used both with preloading and also on a specific instance -- it's just that that scope can't be instance dependent. That's why the documentation says this is possible, you just can't require an argument.

To put that another way: it's useful to be able to define an association scope that _optionally_ does something with an instance but does something differently when not on an instance. For example, you might require that a column be a specific value when an instance is supplied but when an instance is not available only require that that column be not null.

Reply:
@jonathanhefner Did you have any thoughts on my reply?

Reply:
@jcoleman While specific behavior may be useful, I do not think we should document it if it isn't intended to be supported.

As I said, I don't think #36134 intended to support the use case you mentioned.

Furthermore, that use case does not seem to be fully supported due to the difference between lambda and non-lambda Procs:

```console
> ->(x = 1){}.arity # lambda
-1

> Proc.new{|x = 1|}.arity # non-lambda
0
```